File: ios/JejuFolklore/Sources/Models/PlayProgress.swift
import Foundation

/// **이번 판의 상태.** 진행 화면이 살아 있는 동안만 존재한다.
///
/// 화면 세 곳이 이걸 읽는다.
///
///     상단 HUD    생활기록 2 / 6
///     CLEAR       완료 미션 8/8 · 건너뛴 미션 · 걸린 시간
///     진행         지금 어느 Point · 어느 Mission · 몇 번째 Step
///
/// ## `PlayProgressStore` 가 미션 단위로 저장한다 (2026-09-07 되살림)
///
/// 2026-09-04 에 한 번 걷어냈다 — PLAY 는 언제나 처음부터, 화면마다 상태에 따라
/// 다른 말이 나오던 것을 없애 만들고 확인하기를 단순하게 하려고 했다. 그런데
/// 성읍은 60~75분짜리라 현장에서 전화가 오거나 배터리를 아끼려 앱을 내리면
/// 대장간집부터 다시 해야 했다 — 그 손실이 더 크다고 판단해 되살렸다.
///
/// **Step 단위가 아니라 미션 단위로 저장한다.** 미션을 하나 끝낼 때마다
/// `RunnerViewModel` 이 이 구조체를 `PlayProgressStore` 에 write 한다. 그래서
/// 나갔다 들어오면 **마지막으로 완료한 미션 다음**부터 시작하고, 진행 중이던
/// 미션의 Step·힌트 상태는 되살아나지 않는다 — 나가기 확인창이 그 정도를
/// 정확히 말해준다.
///
/// 단, 끝낸 미션의 발견·이야기를 **보던 중**에 나갔으면 그 자리부터 다시 연다
/// (`pendingReveal`, 2026-09-15).
struct PlayProgress: Codable, Equatable {
    let playId: String
    var currentPointId: String
    var currentMissionId: String
    var currentStepIndex: Int
    var completedMissionIds: [String]
    /// 복원한 기록 칸. 「생활기록 4/6」의 4가 이것이다.
    var discoveredRecordIds: [String]
    /// 건너뛴 미션. 진행은 되지만 **직접 발견한 것은 아니다** —
    /// CLEAR 화면에서 구분해 보여주려고 따로 센다.
    var skippedMissionIds: [String]
    var finalCleared: Bool
    /// 아직 다 못 본 발견·이야기. 없으면 nil. 이 칸이 생기기 전(2026-09-15 전)에
    /// 기기에 저장된 진행에는 없다 — 그때는 nil 로 읽는다 (`init(from:)`).
    var pendingReveal: PendingReveal?
    let startedAt: Date
    var updatedAt: Date

    /// 미션은 끝냈지만 **아직 다 못 본** 발견·이야기 (2026-09-15).
    ///
    /// 진행은 미션을 끝내는 순간 저장된다. 그 뒤에 나오는 발견·이야기를 보다가 앱을
    /// 내리면 전에는 이어서 할 때 그걸 건너뛰고 다음 미션(마지막이면 FINAL)으로 갔다 —
    /// 방금 찾은 것의 의미를 못 듣고 지나가는 셈이다. FINAL 이야기를 듣다 나가면 맞힌
    /// FINAL 을 다시 풀어야 했다. 그래서 무엇을 보던 중이었는지 같이 적어 두고,
    /// 이어서 할 때 거기부터 연다. 웹(`miniapp` 의 `PendingReveal`)도 같은 두 경우다.
    enum PendingReveal: Codable, Equatable {
        enum Stage: String, Codable { case discovery, story }
        /// 미션을 끝낸 뒤 그 미션의 발견 또는 이야기를 보던 중.
        case afterMission(missionId: String, stage: Stage)
        /// FINAL 을 맞힌 뒤 FINAL 이야기(`final.story`)를 듣던 중. 끝나면 CLEAR 다.
        case afterFinal
    }

    init(play: Play) {
        let first = play.orderedMissions.first
        self.playId = play.id
        self.currentPointId = first?.point.id ?? ""
        self.currentMissionId = first?.mission.id ?? ""
        self.currentStepIndex = 0
        self.completedMissionIds = []
        self.discoveredRecordIds = []
        self.skippedMissionIds = []
        self.finalCleared = false
        self.pendingReveal = nil
        self.startedAt = Date()
        self.updatedAt = Date()
    }

    /// 저장된 진행을 읽는다. 직접 쓴 이유는 `pendingReveal` 하나다.
    ///
    /// 칸이 없으면(옛 저장) nil 이다. 칸이 있는데 못 읽는 모양이어도(경우가 늘어난
    /// 저장을 옛 앱이 읽을 때) nil 로 본다 — 자동 생성 디코딩은 이때 **진행 전체**를
    /// 못 읽어 `PlayProgressStore` 가 통째로 버린다. 못 읽는 것은 「보다 만 자리」뿐이고,
    /// 그게 없어도 완료한 미션 다음부터 이어갈 수 있다. 쓰기(`encode`)는 자동 생성 그대로다.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        playId = try c.decode(String.self, forKey: .playId)
        currentPointId = try c.decode(String.self, forKey: .currentPointId)
        currentMissionId = try c.decode(String.self, forKey: .currentMissionId)
        currentStepIndex = try c.decode(Int.self, forKey: .currentStepIndex)
        completedMissionIds = try c.decode([String].self, forKey: .completedMissionIds)
        discoveredRecordIds = try c.decode([String].self, forKey: .discoveredRecordIds)
        skippedMissionIds = try c.decode([String].self, forKey: .skippedMissionIds)
        finalCleared = try c.decode(Bool.self, forKey: .finalCleared)
        pendingReveal = try? c.decodeIfPresent(PendingReveal.self, forKey: .pendingReveal)
        startedAt = try c.decode(Date.self, forKey: .startedAt)
        updatedAt = try c.decode(Date.self, forKey: .updatedAt)
    }

    var isFinished: Bool { finalCleared }
}

File: miniapp/src/stores/playProgress.ts
/**
 * PLAY 진행 저장 — Models/PlayProgress.swift + Services/PlayProgressStore.swift 이식.
 *
 * ## 왜 저장하나 (2026-09-07 되살림)
 * 성읍은 60~75분짜리다. 현장에서 전화가 오거나 배터리를 아끼려 앱을 내리면 처음부터
 * 다시 해야 했다. 여러 PLAY 를 동시에 붙잡고 있을 수 있다 — 하다 말고 다른 곳에 갔다가
 * 돌아올 수 있어야 한다.
 *
 * ## 미션 단위로 저장한다
 * 러너가 **미션을 하나 끝낼 때마다** `save()` 한다. 나갔다 들어오면 **마지막으로 완료한
 * 미션 다음**부터 시작하고, 진행 중이던 미션의 Step·힌트는 되살아나지 않는다.
 * (나가기 확인창 문구: 「나가기 (완료한 지점까지 저장돼요)」)
 *
 * ## 다른 곳이 읽는 것
 *   홈 카드      inProgress() → 「이어서 하기」, completed() → 「다시 하기」
 *   PLAY 상세    load(playId) → 버튼 문구, clear(playId) → 「처음부터 다시 하기」
 *   러너         beginPlay() 로 시작/이어받기, 미션 완료·CLEAR 때 save()
 *
 * 저장 키는 iOS 와 같은 `play_progress_v1`. 값은 `{ [playId]: PlayProgress }` JSON.
 * 시각은 epoch **밀리초**(number)로 둔다.
 */
import { orderedMissions } from '../api/models';
import type { Play } from '../api/types';
import { createJsonStore, type JsonStore } from './jsonStore';
import type { KeyValueBackend } from './kv';

export const PLAY_PROGRESS_KEY = 'play_progress_v1';

/** 이번 판의 상태. */
export interface PlayProgress {
  playId: string;
  currentPointId: string;
  currentMissionId: string;
  currentStepIndex: number;
  completedMissionIds: string[];
  /** 복원한 기록 칸. 「생활기록 4/6」의 4. */
  discoveredRecordIds: string[];
  /** 건너뛴 미션. 진행은 되지만 **직접 발견한 것은 아니다** — CLEAR 에서 구분해 보여준다. */
  skippedMissionIds: string[];
  finalCleared: boolean;
  /** epoch ms */
  startedAt: number;
  /** epoch ms. save() 가 채운다. */
  updatedAt: number;
}

export type ProgressMap = Record<string, PlayProgress>;

/** 새 판. 첫 미션에서 시작한다. */
export function createPlayProgress(play: Play, now: number = Date.now()): PlayProgress {
  const first = orderedMissions(play)[0];
  return {
    playId: play.id,
    currentPointId: first?.point.id ?? '',
    currentMissionId: first?.mission.id ?? '',
    currentStepIndex: 0,
    completedMissionIds: [],
    discoveredRecordIds: [],
    skippedMissionIds: [],
    finalCleared: false,
    startedAt: now,
    updatedAt: now,
  };
}

export const isFinished = (p: PlayProgress): boolean => p.finalCleared;

/**
 * 이어받을 미션 번호 (orderedMissions 기준).
 * **완료하지 않은 첫 미션**. 전부 완료했으면 마지막 미션(→ 러너가 FINAL 로 넘긴다).
 */
export function resumeMissionIndex(play: Play, saved: PlayProgress): number {
  const flat = orderedMissions(play);
  const i = flat.findIndex(({ mission }) => !saved.completedMissionIds.includes(mission.id));
  return i >= 0 ? i : Math.max(flat.length - 1, 0);
}

/**
 * 러너가 시작할 때 부른다 (RunnerViewModel.init).
 * - 저장된 진행이 있고 **CLEAR 전**이면 이어받는다 (완료한 미션 다음부터).
 * - 이미 CLEAR 한 기록은 이어받지 않는다 — 「다시 하기」는 처음부터가 맞다.
 */
export function beginPlay(
  play: Play,
  saved: PlayProgress | null,
  now: number = Date.now(),
): { progress: PlayProgress; missionIndex: number; resumed: boolean } {
  if (saved && !isFinished(saved)) {
    return { progress: saved, missionIndex: resumeMissionIndex(play, saved), resumed: true };
  }
  return { progress: createPlayProgress(play, now), missionIndex: 0, resumed: false };
}

/** 진행 지도에 체크를 넣을 Point — 그 Point 의 미션을 **전부** 끝냈을 때. */
export function clearedPointIds(play: Play, progress: PlayProgress): Set<string> {
  return new Set(
    play.points
      .filter((pt) => pt.missions.length > 0 && pt.missions.every((m) => progress.completedMissionIds.includes(m.id)))
      .map((pt) => pt.id),
  );
}

/** 홈 카드·PLAY 상세 버튼 문구. 저장이 없으면 null(→ 「퀘스트 수락」/「플레이하기」는 화면이 정한다). */
export function progressLabel(p: PlayProgress | null | undefined): '이어서 하기' | '다시 하기' | null {
  if (!p) return null;
  return isFinished(p) ? '다시 하기' : '이어서 하기';
}

// ── 저장 형식 검사 ─────────────────────────────────────────────────────

const isStrArr = (v: unknown): v is string[] => Array.isArray(v) && v.every((x) => typeof x === 'string');

function isPlayProgress(v: unknown): v is PlayProgress {
  if (typeof v !== 'object' || v === null) return false;
  const p = v as Record<string, unknown>;
  return (
    typeof p.playId === 'string' &&
    typeof p.currentPointId === 'string' &&
    typeof p.currentMissionId === 'string' &&
    typeof p.currentStepIndex === 'number' &&
    isStrArr(p.completedMissionIds) &&
    isStrArr(p.discoveredRecordIds) &&
    isStrArr(p.skippedMissionIds) &&
    typeof p.finalCleared === 'boolean' &&
    typeof p.startedAt === 'number' &&
    typeof p.updatedAt === 'number'
  );
}

/** 하나라도 못 읽으면 통째로 버린다 — iOS 가 `[String: PlayProgress]` 디코딩 실패 때 그러듯이. */
function parseProgressMap(raw: unknown): ProgressMap | null {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) return null;
  const out: ProgressMap = {};
  for (const [k, v] of Object.entries(raw)) {
    if (!isPlayProgress(v)) return null;
    out[k] = v;
  }
  return out;
}

// ── 저장소 ─────────────────────────────────────────────────────────────

export interface PlayProgressStore {
  readonly raw: JsonStore<ProgressMap>;
  hydrate(): Promise<void>;
  load(playId: string): PlayProgress | null;
  /** updatedAt 을 지금으로 찍어 저장하고, 저장된 값을 돌려준다. */
  save(progress: PlayProgress): PlayProgress;
  clear(playId: string): void;
  /** 진행 중(아직 CLEAR 안 한 것), 최근 것부터. */
  inProgress(): PlayProgress[];
  /** 완료한 것, 최근 것부터. */
  completed(): PlayProgress[];
  all(): ProgressMap;
  subscribe(listener: () => void): () => void;
  whenIdle(): Promise<void>;
}

export function createPlayProgressStore(
  backend: KeyValueBackend,
  opts: { now?: () => number } = {},
): PlayProgressStore {
  const now = opts.now ?? Date.now;
  const raw = createJsonStore<ProgressMap>({ backend, key: PLAY_PROGRESS_KEY, empty: () => ({}), parse: parseProgressMap });
  const byRecent = (a: PlayProgress, b: PlayProgress) => b.updatedAt - a.updatedAt;
  const persist = (next: ProgressMap) => {
    raw.set(next).catch(() => undefined); // 실패는 jsonStore 가 로그로 남긴다. 메모리 값은 유지.
  };

  return {
    raw,
    hydrate: () => raw.hydrate(),
    load: (playId) => raw.get()[playId] ?? null,
    save(progress) {
      const saved: PlayProgress = { ...progress, updatedAt: now() };
      persist({ ...raw.get(), [saved.playId]: saved });
      return saved;
    },
    clear(playId) {
      if (!(playId in raw.get())) return;
      const next = { ...raw.get() };
      delete next[playId];
      persist(next);
    },
    inProgress: () => Object.values(raw.get()).filter((p) => !isFinished(p)).sort(byRecent),
    completed: () => Object.values(raw.get()).filter(isFinished).sort(byRecent),
    all: () => raw.get(),
    subscribe: (l) => raw.subscribe(l),
    whenIdle: () => raw.whenIdle(),
  };
}
